src/examples/zkapps/dex/upgradability.ts

```diff
@@ -0,0 +1,456 @@
+    feePayerUpdate.send({ to: addresses.tokenY, amount: accountFee.mul(2) });
+    tokenX.deploy();
+    tokenY.deploy();
+  });
+  await tx.prove();
+  tx.sign([keys.tokenX, keys.tokenY]);
+  await tx.send();
+  balances = getTokenBalances();
+  console.log(
+    'Token contract tokens (X, Y):',
+    balances.tokenContract.X,
+    balances.tokenContract.Y
+  );
+
+  /**
+   * # Atomic Actions 1
+   *
+   * Preconditions:
+   *  - SC deployed with a non-writable delegate field (impossible permission)
+   *
+   * Actions:
+   *  - User sets permissions for delegate to be edible
+   *  - User changes delegate field to something
+   *
+   * Expected:
+   *  - delegate now holds a new address
+   */
+
+  console.log('deploy dex contracts...');
+  tx = await Mina.transaction(feePayerKey, () => {
+    // pay fees for creating 3 dex accounts
+    AccountUpdate.createSigned(feePayerKey).balance.subInPlace(
+      accountFee.mul(3)
+    );
+    dex.deploy();
+    tokenX.deployZkapp(addresses.dex, DexTokenHolder._verificationKey!);
+    tokenY.deployZkapp(addresses.dex, DexTokenHolder._verificationKey!);
+
+    // setting the setDelegate permission field to impossible
+    let dexAccount = AccountUpdate.create(addresses.dex);
+    AccountUpdate.setValue(dexAccount.update.permissions, {
+      ...Permissions.initial(),
+      setDelegate: Permissions.impossible(),
+    });
+    dexAccount.sign();
+  });
+  await tx.prove();
+  tx.sign([keys.dex]);
+  await tx.send();
+
+  console.log(
+    'trying to change delegate (setDelegate=impossible, should fail)'
+  );
+  tx = await Mina.transaction(feePayerKey, () => {
+    // setting the delegate field to something, although permissions forbid it
+    let dexAccount = AccountUpdate.create(addresses.dex);
+    AccountUpdate.setValue(
+      dexAccount.update.delegate,
+      PrivateKey.random().toPublicKey()
+    );
+    dexAccount.sign();
+  });
+  await tx.prove();
+
+  await expect(tx.sign([keys.dex]).send()).rejects.toThrow(
+    /Cannot update field 'delegate'/
+  );
+
+  console.log('changing delegate permission back to normal');
+
+  tx = await Mina.transaction(feePayerKey, () => {
+    let dexAccount = AccountUpdate.create(addresses.dex);
+    AccountUpdate.setValue(dexAccount.update.permissions, {
+      ...Permissions.initial(),
+      setDelegate: Permissions.proofOrSignature(),
+    });
+    dexAccount.sign();
+  });
+  await tx.prove();
+  await tx.sign([keys.dex]).send();
+
+  console.log('changing delegate field to a new address');
+
+  let newDelegate = PrivateKey.random().toPublicKey();
+  tx = await Mina.transaction(feePayerKey, () => {
+    let dexAccount = AccountUpdate.create(addresses.dex);
+    AccountUpdate.setValue(dexAccount.update.delegate, newDelegate);
+    dexAccount.sign();
+  });
+  await tx.prove();
+  await tx.sign([keys.dex]).send();
+
+  Mina.getAccount(addresses.dex).delegate?.assertEquals(newDelegate);
```

Comment:
Would calls to smart contract methods fail at this point since the delegate is set to a random address? It's not a problem (delegates are exercised in the upgradability section), just checking that I understand correctly.

Reply:
I think in that case it should be fine and not fail